src/Features/Protocol/LanguageServices/RoslynLanguageService.cs

```diff
@@ -0,0 +1,640 @@
+        public Task<VSLocation[]> GoToDefinitionAsync(Solution solution, TextDocumentPositionParams request, CancellationToken cancellationToken)
+        {
+            return GetDefinitionAsync(solution, request, typeOnly: false, cancellationToken);
+        }
+
+        /// <summary>
+        /// Returns the location of the definition for a given symbol
+        /// </summary>
+        /// <returns>The location of the definition of the given symbol</returns>
+        public Task<VSLocation[]> GoToTypeDefinitionAsync(Solution solution, TextDocumentPositionParams request, CancellationToken cancellationToken)
+        {
+            return GetDefinitionAsync(solution, request, typeOnly: true, cancellationToken);
+        }
+
+        /// <summary>
+        /// Finds all references for a given symbol
+        /// </summary>
+        /// <returns>Returns a list of locations of references to the given symbol</returns>
+        public async Task<VSLocation[]> FindAllReferencesAsync(Solution solution, ReferenceParams request, CancellationToken cancellationToken)
+        {
+            var locations = new List<VSLocation>();
+
+            var document = solution?.GetDocument(request.TextDocument.Uri);
+            if (document == null)
+            {
+                return locations.ToArray();
+            }
+
+            var findUsagesService = document.Project.LanguageServices.GetService<IFindUsagesService>();
+            var position = await document.GetPositionAsync(request.Position.ToLinePosition(), cancellationToken).ConfigureAwait(false);
+
+            var context = new SimpleFindUsagesContext(cancellationToken);
+
+            await findUsagesService.FindReferencesAsync(document, position, context).ConfigureAwait(false);
+
+            if (request.Context.IncludeDeclaration)
+            {
+                foreach (var definition in context.GetDefinitions())
+                {
+                    foreach (var docSpan in definition.SourceSpans)
+                    {
+                        locations.Add(await docSpan.ToLocationAsync(cancellationToken).ConfigureAwait(false));
+                    }
+                }
+            }
+
+            foreach (var reference in context.GetReferences())
+            {
+                locations.Add(await reference.SourceSpan.ToLocationAsync(cancellationToken).ConfigureAwait(false));
+            }
+
+            return locations.ToArray();
+        }
+
+        /// <summary>
+        /// Finds all implementation for a given symbol
+        /// </summary>
+        /// <returns>Returns a list of locations of references to the given symbol</returns>
+        public async Task<VSLocation[]> FindAllImplementationsAsync(Solution solution, TextDocumentPositionParams request, CancellationToken cancellationToken)
+        {
+            var locations = new List<VSLocation>();
+
+            var document = solution?.GetDocument(request.TextDocument.Uri);
```

Comment:
I think we should just assume solution is always given rather than handle solution being null case now